SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityDeployChecker.java

```diff
@@ -370,74 +386,211 @@ private SingularityDeployResult getDeployResult(final SingularityRequest request
+                  new SingularityLoadBalancerUpdate(BaragonRequestState.UNKNOWN, lbRequestId, Optional.<String> absent(), System.currentTimeMillis(), LoadBalancerMethod.PRE_ENQUEUE,
+                    Optional.<String> absent()));
+                SingularityLoadBalancerUpdate enqueueResult = lbClient.enqueue(lbRequestId, request, deploy.get(), getTasks(deployActiveTasks, tasks), getTasks(toShutDown, tasks));
+                updateLoadBalancerStateForTasks(deployActiveTasks, LoadBalancerRequestType.ADD, enqueueResult);
+
+                DeployState deployState = interpretLoadBalancerState(enqueueResult, DeployState.WAITING);
+                if (deployState == DeployState.SUCCEEDED) {
+                  updatePendingDeploy(pendingDeploy, Optional.of(enqueueResult), DeployState.WAITING);
+                  return markStepFinished(pendingDeploy, deploy, otherActiveTasks, request, updatePendingDeployRequest);
+                } else if (deployState == DeployState.WAITING) {
+                  updatePendingDeploy(pendingDeploy, Optional.of(enqueueResult), deployState);
+                  return new SingularityDeployResult(DeployState.WAITING);
+                } else {
+                  return fromLbState(deployState, enqueueResult);
+                }
+              }
+            } else {
+              return markStepFinished(pendingDeploy, deploy, otherActiveTasks, request, updatePendingDeployRequest);
+            }
+          case UNHEALTHY:
+          default:
+            return new SingularityDeployResult(DeployState.FAILED, "Not all tasks for deploy were healthy");
+        }
+      } else {
+        maybeUpdatePendingRequest(pendingDeploy, deploy, request, updatePendingDeployRequest);
+        return checkOverdue(deploy, deployActiveTasks, isDeployOverdue);
       }
     }
+  }
+
+  private void maybeUpdatePendingRequest(SingularityPendingDeploy pendingDeploy, Optional<SingularityDeploy> deploy, SingularityRequest request, Optional<SingularityUpdatePendingDeployRequest> updatePendingDeployRequest) {
+    maybeUpdatePendingRequest(pendingDeploy, deploy, request, updatePendingDeployRequest, Optional.<SingularityLoadBalancerUpdate> absent());
+  }
 
-    if (pendingDeploy.getLastLoadBalancerUpdate().isPresent()) {
-      return new SingularityDeployResult(DeployState.WAITING, Optional.of("Waiting on load balancer API"), pendingDeploy.getLastLoadBalancerUpdate(), System.currentTimeMillis());
+  private void maybeUpdatePendingRequest(SingularityPendingDeploy pendingDeploy, Optional<SingularityDeploy> deploy, SingularityRequest request, Optional<SingularityUpdatePendingDeployRequest> updatePendingDeployRequest, Optional<SingularityLoadBalancerUpdate> lbUpdate) {
+    if (updatePendingDeployRequest.isPresent() && pendingDeploy.getDeployProgress().isPresent()) {
+      SingularityDeployProgress deployProgress = pendingDeploy.getDeployProgress().get();
+      SingularityDeployProgress newProgress = new SingularityDeployProgress(
```

Comment:
Like deployProgress.transform/progress/nextStep/clone(pass in changed parameters)

Not sure what the right name is.